File: app/src/main/java/ru/magflayer/spectrum/domain/model/ListType.java
package ru.magflayer.spectrum.domain.model;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.List;

public final class ListType<T> implements ParameterizedType {
    private Class<?> wrapped;

    public ListType(Class<T> wrapper) {
        this.wrapped = wrapper;
    }

    @Override
    public Type[] getActualTypeArguments() {
        return new Type[]{wrapped};
    }

    @Override
    public Type getRawType() {
        return List.class;
    }

    @Override
    public Type getOwnerType() {
        return null;
    }
}

File: app/src/main/java/ru/magflayer/spectrum/domain/entity/ListType.kt
package ru.magflayer.spectrum.domain.entity

import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type

class ListType<T>(wrapper: Class<T>) : ParameterizedType {
    private val wrapped: Class<*>

    init {
        this.wrapped = wrapper
    }

    override fun getActualTypeArguments(): Array<Type> {
        return arrayOf(wrapped)
    }

    override fun getRawType(): Type {
        return List::class.java
    }

    override fun getOwnerType(): Type? {
        return null
    }
}
